`emmylua_doc_cli`: Compilation error on 0.7.1 resulting in failed build. Something is wrong with cargo versioning.

https://github.com/EmmyLuaLs/emmylua-analyzer-rust/blob/07c845b8cd95b4336219d750721edbe38ae1e968/crates/emmylua_doc_cli/Cargo.toml#L3

https://crates.io/crates/emmylua_doc_cli

![Image](https://github.com/user-attachments/assets/5eed54d4-1d59-41c4-adf2-55474723a6e2)

I don't understand how/why this is the case.

We just had a compilation error in our site build workflow: https://github.com/beyond-all-reason/RecoilEngine/actions/runs/14940110310/job/41975763861

This happened on version 0.7.1: https://github.com/beyond-all-reason/RecoilEngine/blob/8bacfa481d8b63cf1a62733adc6c934ef4818efa/.github/workflows/publish-site.yml#L7

We have been successfully using 0.7.1 and it suddenly failed to compile. We suspect that there is a problem with dependency between `emmylua_doc_cli` and `emmylua_code_analysis`

```
   Compiling emmylua_code_analysis v0.7.2
   Compiling emmylua_doc_cli v0.7.1
error[E0609]: no field `is_deprecated` on type `&emmylua_code_analysis::db_index::property::property::LuaCommonProperty`
  --> /home/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/emmylua_doc_cli-0.7.1/src/markdown_generator/gen/mod.rs:26:21
   |
26 |         if property.is_deprecated {
   |                     ^^^^^^^^^^^^^ unknown field
   |
help: a field with a similar name exists
   |
26 -         if property.is_deprecated {
26 +         if property.deprecated {
   |

error[E0609]: no field `deprecated_message` on type `&emmylua_code_analysis::db_index::property::property::LuaCommonProperty`
  --> /home/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/emmylua_doc_cli-0.7.1/src/markdown_generator/gen/mod.rs:27:48
   |
27 |             if let Some(deprecated) = property.deprecated_message.clone() {
   |                                                ^^^^^^^^^^^^^^^^^^ unknown field
   |
   = note: available fields are: `id`, `description`, `visibility`, `source`, `deprecated` ... and 3 others

error[E0599]: no method named `iter` found for reference `&LuaType` in the current scope
  --> /home/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/emmylua_doc_cli-0.7.1/src/markdown_generator/render.rs:74:10
   |
73 |       let ret_strs = rets
   |  ____________________-
74 | |         .iter()
   | |         -^^^^ method not found in `&LuaType`
   | |_________|
   |

Some errors have detailed explanations: E0599, E0609.
For more information about an error, try `rustc --explain E0599`.
error: could not compile `emmylua_doc_cli` (bin "emmylua_doc_cli") due to 3 previous errors
error: failed to compile `emmylua_doc_cli v0.7.1`, intermediate artifacts can be found at `/tmp/cargo-install5Js3nP`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
```

--- a/.github/workflows/bump-upload.yml
+++ b/.github/workflows/bump-upload.yml
@@ -32,7 +32,7 @@ jobs:
           file_pattern: 'crates/*/Cargo.toml'
           
   upload-crates:
-    uses: ./.github/workflows/upload-crates.yml
+    uses: ./.github/workflows/upload_crates.yml
 
   upload-cachix:
-    uses: ./.github/workflows/upload-cachix.yml
+    uses: ./.github/workflows/upload_cachix.yml

--- a/.github/workflows/bump-upload.yml
+++ b/.github/workflows/bump-upload.yml
@@ -1,0 +1,37 @@
+name: Bump Version & Upload
+
+on:
+  push:
+    tags:
+      - '*'
+
+jobs:
+  bump:
+    name: Bump version
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write
+    steps:
+      - name: Clone repo
+        uses: actions/checkout@v4
+
+      - name: Install Rust toolchain
+        uses: dtolnay/rust-toolchain@stable
+
+      - id: version
+        name: Bump version
+        run: |
+          echo "current ref ${{ github.ref }}"
+          cargo run -p edit_version -- ${{ github.ref }}
+          echo "VERSION=$(sed 's:refs/tags/::' <<< ${{ github.ref }})" >> "$GITHUB_OUTPUT"
+
+      - name: Push changes
+        uses: stefanzweifel/git-auto-commit-action@v5
+        with:
+          commit_message: Bump version to ${{ steps.version.outputs.VERSION }}
+          file_pattern: 'crates/*/Cargo.toml'
+          
+  upload-crates:
+    uses: ./.github/workflows/upload-crates.yml
+  upload-cachix:
+    uses: ./.github/workflows/upload-cachix.yml

--- a/.github/workflows/bump-upload.yml
+++ b/.github/workflows/bump-upload.yml
@@ -33,5 +33,6 @@ jobs:
           
   upload-crates:
     uses: ./.github/workflows/upload-crates.yml
+
   upload-cachix:
     uses: ./.github/workflows/upload-cachix.yml

--- a/.github/workflows/upload_cachix.yml
+++ b/.github/workflows/upload_cachix.yml
@@ -1,10 +1,8 @@
 name: "Upload to Cachix"
 
 on:
-  push:
-    tags:
-      - '*'
   workflow_dispatch:
+  workflow_call:
 
 jobs:
   build-and-cache:
@@ -18,14 +16,8 @@ jobs:
           
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v4
-      with:
-        fetch-depth: 0
-      
-    - name: Set version
-      run: |
-        echo "current ref ${{ github.ref }}"
-        cargo run -p edit_version -- ${{ github.ref }}
+    - name: Clone repo
+      uses: actions/checkout@v4
       
     - name: Setup Nix
       uses: cachix/install-nix-action@v30
@@ -47,4 +39,7 @@ jobs:
         
     - name: Build emmylua-analyzer
       run: |
-        nix build .#emmylua_{ls,doc_cli,check}
+        for x in emmylua_{ls,doc_cli,check}; do
+          nix build --impure --expr \
+            "with import <nixpkgs> {}; callPackage (import ./nix/packages.nix).$x {}"
+        done

--- a/.github/workflows/upload_crates.yml
+++ b/.github/workflows/upload_crates.yml
@@ -21,7 +21,7 @@ jobs:
       - name: Publish
         run: |
           for dir in emmylua_{code_analysis,ls,doc_cli,check}; do
-            pushd "crates/$x"
+            pushd "crates/$dir"
             cargo publish --allow-dirty
             popd
           done

--- a/.github/workflows/upload_crates.yml
+++ b/.github/workflows/upload_crates.yml
@@ -1,17 +1,15 @@
 name: Upload Crates
 
 on:
-  push:
-    tags:
-      - '*'
+  workflow_call:
 
 jobs:
   upload:
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v4
-        with:
-          submodules: recursive
+      - name: Clone repo
+        uses: actions/checkout@v4
+
       - name: Install Rust toolchain
         uses: dtolnay/rust-toolchain@stable
 
@@ -19,15 +17,13 @@ jobs:
         run: |
           echo "current ref ${{ github.ref }}"
           cargo run -p edit_version -- ${{ github.ref }}
+
       - name: Publish
         run: |
-          cd crates/emmylua_code_analysis
-          cargo publish --allow-dirty
-          cd ../emmylua_ls
-          cargo publish --allow-dirty
-          cd ../emmylua_doc_cli
-          cargo publish --allow-dirty
-          cd ../emmylua_check
-          cargo publish --allow-dirty
+          for dir in emmylua_{code_analysis,ls,doc_cli,check}; do
+            pushd "crates/$x"
+            cargo publish --allow-dirty
+            popd
+          done
         env:
           CARGO_REGISTRY_TOKEN: ${{ secrets.CARGO_TOKEN }}

--- a/flake.nix
+++ b/flake.nix
@@ -3,58 +3,35 @@
 
   inputs = {
     nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
+    flake-utils.url = "github:numtide/flake-utils";
     rust-overlay = {
       url = "github:oxalica/rust-overlay";
       inputs.nixpkgs.follows = "nixpkgs";
     };
-    flake-utils.url = "github:numtide/flake-utils";
   };
 
   outputs =
     {
+      self,
       nixpkgs,
-      rust-overlay,
       flake-utils,
+      rust-overlay,
       ...
     }:
     flake-utils.lib.eachDefaultSystem (
       system:
       let
-        overlays = [ (import rust-overlay) ];
+        overlays = [ rust-overlay.overlays.default ];
         pkgs = import nixpkgs { inherit system overlays; };
-
-        inherit (pkgs) rustPlatform;
       in
       {
         packages =
           let
-            mkPackage =
-              x:
-              let
-                cargoToml = builtins.fromTOML (builtins.readFile ./crates/${x}/Cargo.toml);
-              in
-              rustPlatform.buildRustPackage {
-                pname = cargoToml.package.name;
-                version = cargoToml.package.version;
-
-                src = ./.;
-                cargoLock.lockFile = ./Cargo.lock;
-
-                nativeBuildInputs = with pkgs; [ pkg-config ];
-                buildInputs = with pkgs; [ stdenv.cc.cc.lib ];
-
-                buildAndTestSubdir = "crates/${x}";
-
-                postFixup = ''
-                  patchelf --set-rpath "${pkgs.stdenv.cc.cc.lib}/lib" $out/bin/${x}
-                '';
-              };
+            packages = import ./nix/packages.nix;
           in
-          rec {
-            emmylua_ls = mkPackage "emmylua_ls";
-            emmylua_doc_cli = mkPackage "emmylua_doc_cli";
-            emmylua_check = mkPackage "emmylua_check";
-            default = emmylua_ls;
+          (builtins.mapAttrs (name: value: pkgs.callPackage value { }) packages)
+          // {
+            default = self.packages.${system}.emmylua_ls;
           };
       }
     );

--- a/nix/packages.nix
+++ b/nix/packages.nix
@@ -1,0 +1,47 @@
+let
+  root = ../.;
+  mkPackage =
+    x:
+    {
+      rustPlatform,
+      stdenv,
+      pkg-config,
+    }:
+    let
+      cargoToml = builtins.fromTOML (builtins.readFile /${root}/crates/${x}/Cargo.toml);
+    in
+    rustPlatform.buildRustPackage {
+      pname = cargoToml.package.name;
+      version = cargoToml.package.version;
+
+      src = root;
+      cargoLock.lockFile = root + /Cargo.lock;
+
+      nativeBuildInputs = [ pkg-config ];
+      buildInputs = [ stdenv.cc.cc.lib ];
+
+      buildAndTestSubdir = "crates/${x}";
+
+      postFixup = ''
+        patchelf --set-rpath "${stdenv.cc.cc.lib}/lib" $out/bin/${x}
+      '';
+    };
+in
+(builtins.listToAttrs (
+  map
+    (
+      x:
+      let
+        name = "emmylua_${x}";
+      in
+      {
+        inherit name;
+        value = mkPackage name;
+      }
+    )
+    [
+      "ls"
+      "doc_cli"
+      "check"
+    ]
+))

